pallets/utxo/src/rewards.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2021 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://spdx.org/licenses/MIT
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author(s): C. Yap
+
+use crate::{Config, Pallet, Event, BlockAuthorRewardAmount, BlockAuthor, MLTCoinsAvailable, TransactionOutput, UtxoStore, StartingPeriod};
+
+use frame_support::traits::Get;
+use sp_runtime::traits::{SaturatedConversion, Saturating, BlakeTwo256, Hash, Zero};
+
+
+/// Returns `true` if reward reduction period has passed; and if a new period has to be created.
+pub(crate) fn period_elapsed<T:Config>(time_now: T::BlockNumber) -> bool {
+    let starting_period =  <StartingPeriod<T>>::get();
+    let time_elapsed = time_now.saturating_sub(starting_period);
+    let has_elapse = time_elapsed > T::RewardReductionPeriod::get();
+
+    if has_elapse {
+        log::debug!("period has elapsed. Updating the start of period to now");
+        <StartingPeriod<T>>::put(time_now);
+    }
+    has_elapse
+}
+
+/// Returns the newly reduced reward amount for a Block Author.
+/// How much a reward is reduced, is based on the config's`RewardReductionFraction`.
+pub(super) fn update_reward_amount<T:Config>(coins_available:u128) -> u128 {
+    let reduction_fraction = T::RewardReductionFraction::get();
+
+    let reward_amount = <BlockAuthorRewardAmount<T>>::take();
+    log::info!("current reward amount: {}", reward_amount);
+    let reward_amount_deducted = reduction_fraction.mul_ceil(reward_amount);
+    if let Some(reward_amount) = reward_amount.checked_sub(reward_amount_deducted) {
+
+        // as long as there is still coins available, set the reward to 1.
+        return if reward_amount.is_zero() {
```

Comment:
Where does this check coins_available > 0 ? Have I missed something really obvious?

Reply:
Doesn't this mean the block reward never drops to 0?